File: mini-newtonian/src/constraints.jl
# constraints.jl — 约束系统：滑轮、斜面、张力
# 参考：MIT 8.012 Ch.7, Kleppner & Kolenkow Ch.3

## ============================================================
## 斜面
## ============================================================

"""
斜面上物体的加速度（沿斜面方向）
theta: 斜面倾角
mu: 动摩擦因数（0 = 无摩擦）
返回沿斜面向下的加速度大小
"""
function incline_acceleration(theta::Float64; mu::Float64=0.0, g::Float64=G_EARTH)
    return g * (sin(theta) - mu * cos(theta))
end

"""
斜面上物体不下滑的临界角（静摩擦）
mu_s: 静摩擦因数
"""
function angle_of_repose(mu_s::Float64)
    return atan(mu_s)
end

"""
斜面上的法向力
"""
function normal_force_incline(m::Float64, theta::Float64; g::Float64=G_EARTH)
    return m * g * cos(theta)
end

"""
斜面上物体的运动（数值模拟）
返回沿斜面方向的加速度矢量（3D，斜面在 xz 平面，沿 +x 向上）
"""
function incline_accel_3d(r::Vec3, v::Vec3, t::Float64, theta::Float64; mu::Float64=0.0, g::Float64=G_EARTH)
    # 斜面法向：(-sinθ, 0, cosθ)，沿斜面向下：(cosθ, 0, sinθ)
    g_vec = Vec3(0.0, 0.0, -g)
    normal = Vec3(-sin(theta), 0.0, cos(theta))
    # 法向分量被约束力抵消
    g_normal = dot(g_vec, normal) * normal
    g_tangential = g_vec - g_normal
    # 摩擦力沿速度反方向
    friction = Vec3()
    spd = norm(v)
    if spd > 1e-10
        friction = -mu * m * abs(dot(g_vec, normal)) * (v / spd)
    end
    return g_tangential + friction
end

## ============================================================
## 滑轮系统
## ============================================================

"""
Atwood 机（两个质量通过轻绳跨过无摩擦滑轮）
返回 (加速度 a₁, 加速度 a₂, 绳子张力 T)
a₁ > 0 表示 m₁ 向下加速
"""
function atwood_machine(m1::Float64, m2::Float64; g::Float64=G_EARTH)
    total_mass = m1 + m2
    a = (m1 - m2) * g / total_mass  # m1 向下为正
    T = 2 * m1 * m2 * g / total_mass
    return a, -a, T  # a1 = a, a2 = -a
end

"""
多滑轮系统（n 段绳，各段力相等）
动滑轮组：n 段绳 → 拉力放大 n 倍
"""
function pulley_system_mechanical_advantage(n_segments::Int, load_mass::Float64; g::Float64=G_EARTH)
    effort = load_mass * g / n_segments
    return effort
end

"""
滑轮系统的加速度（给定拉力 F_pull）
负载质量 m_load，拉力 F_pull，滑轮组绳段数 n
返回负载向上的加速度
"""
function pulley_acceleration(m_load::Float64, F_pull::Float64, n_segments::Int=1; g::Float64=G_EARTH)
    F_net = n_segments * F_pull - m_load * g
    return F_net / m_load
end

## ============================================================
## 张力分析
## ============================================================

"""
轻绳中的张力（质量忽略不计）
两端受力相等，张力沿绳方向
"""
function rope_tension(attached_mass::Float64, acceleration::Vec3, other_forces::Vec3=Vec3(); g::Float64=G_EARTH)
    # T + other_forces + mg = ma → T = ma - mg - other_forces
    g_vec = Vec3(0.0, 0.0, -g)
    return attached_mass * acceleration - attached_mass * g_vec - other_forces
end

"""
圆锥摆（conical pendulum）
绳长 L，摆锤质量 m，角速度 omega
返回 (绳与竖直方向夹角, 张力大小)
"""
function conical_pendulum_parameters(m::Float64, L::Float64, omega::Float64; g::Float64=G_EARTH)
    # T*cosθ = mg, T*sinθ = m*ω²*L*sinθ
    # → cosθ = g/(ω²*L)
    cos_theta = g / (omega^2 * L)
    if cos_theta > 1.0 || cos_theta < 0.0
        return NaN, NaN  # 角速度不够大
    end
    theta = acos(cos_theta)
    T = m * g / cos_theta
    return theta, T
end

"""
圆锥摆稳定条件：ω > √(g/L)
"""
function conical_pendulum_critical_omega(L::Float64; g::Float64=G_EARTH)
    return sqrt(g / L)
end

## ============================================================
## 曲面约束
## ============================================================

"""
质点在光滑曲面上运动的法向约束力
曲率半径 R，速度 v，质量 m
"""
function surface_normal_force(m::Float64, v::Float64, R::Float64, theta::Float64; g::Float64=G_EARTH)
    # N = mg*cosθ + mv²/R（向心加速度贡献）
    return m * (g * cos(theta) + v^2 / R)
end

"""
物体脱离曲面的条件：N ≤ 0
"""
function detachment_condition(m::Float64, v::Float64, R::Float64, theta::Float64; g::Float64=G_EARTH)
    N = surface_normal_force(m, v, R, theta; g=g)
    return N <= 0.0, N
end

"""
过山车/环形轨道最低速度（顶点）
R: 环半径
"""
function loop_the_loop_min_speed(R::Float64; g::Float64=G_EARTH)
    return sqrt(g * R)  # v²/R = g → 向心力 = 重力
end

## ============================================================
## 斜面 + 滑轮组合（经典习题）
## ============================================================

"""
斜面上的物体通过滑轮连接悬挂物
m1 在倾角 theta 的斜面上（摩擦因数 mu），通过轻绳跨过滑轮连接 m2（竖直悬挂）
返回 (系统加速度, 绳子张力), a>0 表示 m1 沿斜面向上
"""
function incline_pulley_system(m1::Float64, m2::Float64, theta::Float64; mu::Float64=0.0, g::Float64=G_EARTH)
    total = m1 + m2
    # 沿斜面方向：T - m1*g*sinθ - μ*m1*g*cosθ = m1*a
    # 竖直方向：m2*g - T = m2*a
    # → a = (m2*g - m1*g*sinθ - μ*m1*g*cosθ) / (m1+m2)
    a = (m2 - m1*sin(theta) - mu*m1*cos(theta)) * g / total
    T = m2 * (g - a)
    return a, T
end

"""
判断斜面-滑轮系统的运动方向
返回 :m1_up（m1 沿斜面向上）, :m1_down（m1 沿斜面向下）, :static（静止）
"""
function incline_pulley_direction(m1::Float64, m2::Float64, theta::Float64; mu_s::Float64=0.0, g::Float64=G_EARTH)
    # 无摩擦时纯重力比较
    force_m2 = m2 * g
    force_m1_down = m1 * g * sin(theta)

    if abs(force_m2 - force_m1_down) <= mu_s * m1 * g * cos(theta)
        return :static
    elseif force_m2 > force_m1_down
        return :m1_up
    else
        return :m1_down
    end
end

## ============================================================
## 非惯性系中的有效重力
## ============================================================

"""
加速参考系中的有效重力
a_frame: 参考系的加速度
"""
function effective_gravity(a_frame::Vec3; g::Float64=G_EARTH, g_direction::Vec3=Vec3(0,0,-1))
    g_real = g * normalize(g_direction)
    return g_real - a_frame
end

"""
加速上升/下降的电梯中的表观重量
a_elevator > 0 表示向上加速
"""
function apparent_weight_in_elevator(m::Float64, a_elevator::Float64; g::Float64=G_EARTH)
    return m * (g + a_elevator)  # a_elevator>0上升→更重; a_elevator<0下降→更轻
end

"""
车辆在弯道上的最佳倾斜角（无侧向摩擦）
v: 车速, R: 弯道半径
"""
function banked_curve_angle(v::Float64, R::Float64; g::Float64=G_EARTH)
    tan_theta = v^2 / (g * R)
    return atan(tan_theta)
end

"""
弯道上车辆的最大安全速度（给定倾斜角 theta 和摩擦系数 mu）
"""
function max_safe_speed_banked_curve(R::Float64, theta::Float64, mu::Float64; g::Float64=G_EARTH)
    # v_max = √(g*R * (tanθ + μ) / (1 - μ*tanθ))
    tan_theta = tan(theta)
    numerator = tan_theta + mu
    denominator = 1.0 - mu * tan_theta
    if denominator <= 0
        return Inf  # 不会滑出
    end
    return sqrt(g * R * numerator / denominator)
end

File: mini-newtonian/src/constraints.c
/**
 * constraints.c - Constrained mechanical systems
 * Each function implements an independent knowledge point.
 * Course: MIT 8.012 Ch.7, Kleppner & Kolenkow Ch.3, Goldstein Ch.1.3
 */
#include "constraints.h"
#include <math.h>
#include <stdlib.h>

/* ================================================================
 * L6: Inclined plane
 *
 * Knowledge point: Decomposing gravity into components parallel
 * and perpendicular to the incline. Friction opposes motion.
 *
 * a_parallel = g*(sin(theta) - mu*cos(theta))
 * N = mg*cos(theta)
 * Angle of repose: theta_c = arctan(mu_s)
 * ================================================================ */

double incline_acceleration(double theta, double mu, double g) {
    return g * (sin(theta) - mu * cos(theta));
}
double incline_normal_force(double m, double theta, double g) {
    return m * g * cos(theta);
}
double angle_of_repose(double mu_s) {
    return atan(mu_s);
}
double incline_time_to_bottom(double length, double theta, double mu, double g) {
    double a = incline_acceleration(theta, mu, g);
    if (a <= 0.0) return -1.0;  /* no motion or slides up */
    return sqrt(2.0 * length / a);
}
double incline_velocity_at_bottom(double length, double theta, double mu, double g) {
    double a = incline_acceleration(theta, mu, g);
    if (a <= 0.0) return 0.0;
    return sqrt(2.0 * a * length);
}

/* ================================================================
 * L6: Pulley systems
 *
 * Knowledge point: Ideal pulley redirects tension force.
 * Massless, frictionless string: tension constant throughout.
 *
 * Atwood machine:
 *   a = (m1 - m2)*g / (m1 + m2)  [m1 down positive]
 *   T = 2*m1*m2*g / (m1 + m2)
 *
 * Mechanical advantage: n rope segments -> force multiplied by n
 * ================================================================ */

void atwood_machine(double m1, double m2, double g,
                    double *a1, double *a2, double *tension) {
    double M = m1 + m2;
    double a = (m1 - m2) * g / M;  /* positive = m1 down */
    *a1 = a;
    *a2 = -a;
    *tension = 2.0 * m1 * m2 * g / M;
}
double pulley_mechanical_advantage(int n_segments, double load_mass, double g) {
    return load_mass * g / n_segments;  /* effort needed */
}
double pulley_system_acceleration(double m_load, double F_pull,
                                  int n_segments, double g) {
    double F_net = n_segments * F_pull - m_load * g;
    return F_net / m_load;
}
void atwood_with_friction(double m1, double m2, double theta,
                          double mu, double g, double *a, double *T) {
    /* m1 on incline (angle theta, friction mu), m2 hanging vertically */
    double M = m1 + m2;
    double f1_parallel = m1 * g * sin(theta);
    double f1_friction = mu * m1 * g * cos(theta);
    double a_val = (m2 * g - f1_parallel - f1_friction) / M;
    *a = a_val;
    *T = m2 * (g - a_val);
}

/* ================================================================
 * L6: Tension analysis
 * ================================================================ */

Vec3 rope_tension_force(Vec3 attach_point, Vec3 anchor_point, double tension_magnitude) {
    Vec3 direction = vec3_normalize(vec3_sub(anchor_point, attach_point));
    return vec3_scale(direction, tension_magnitude);
}
double tension_in_accelerating_rope(double attached_mass, double acceleration,
                                    double other_force_magnitude, double g) {
    return attached_mass * (g + acceleration) + other_force_magnitude;
}

/* ================================================================
 * L6: Conical pendulum
 *
 * Knowledge point: Circular motion with a string constraint.
 * T*cos(theta) = mg  (vertical balance)
 * T*sin(theta) = m*omega^2*L*sin(theta) = m*v^2/(L*sin(theta))
 * -> cos(theta) = g/(omega^2*L)
 *
 * Critical omega: omega_min = sqrt(g/L)
 * ================================================================ */

int conical_pendulum_parameters(double m, double L, double omega, double g,
                                double *theta, double *tension) {
    double omega2L = omega * omega * L;
    if (omega2L < g) return 0;  /* omega too small, not a cone */
    double cos_theta = g / omega2L;
    if (cos_theta > 1.0 || cos_theta <= 0.0) return 0;
    *theta = acos(cos_theta);
    *tension = m * g / cos_theta;
    return 1;
}
double conical_pendulum_critical_omega(double L, double g) {
    return sqrt(g / L);
}
double conical_pendulum_radius_at_omega(double L, double omega, double g) {
    double omega2L = omega * omega * L;
    if (omega2L < g) return 0.0;
    double cos_theta = g / omega2L;
    return L * sqrt(1.0 - cos_theta * cos_theta);
}

/* ================================================================
 * L6: Curved surface constraints
 *
 * Knowledge point: Normal force on a curved surface has two contributions:
 *   1. mg*cos(theta) from gravity component perpendicular to surface
 *   2. m*v^2/R from centripetal acceleration requirement
 *
 * Detachment condition: N <= 0 (surface can only push, not pull)
 *
 * Loop-the-loop:
 *   At top: N + mg = mv^2/R -> N = m(v^2/R - g)
 *   Minimum speed at top: v_min = sqrt(gR)
 *   At bottom: N - mg = mv^2/R -> N = m(v^2/R + g)
 * ================================================================ */

double surface_normal_force(double m, double v, double R, double theta, double g) {
    return m * (g * cos(theta) + v * v / R);
}
int detachment_condition(double m, double v, double R, double theta, double g,
                         double *normal_force) {
    *normal_force = surface_normal_force(m, v, R, theta, g);
    return (*normal_force <= 0.0) ? 1 : 0;
}
double loop_the_loop_min_speed(double R, double g) {
    return sqrt(g * R);  /* at the top */
}
double loop_the_loop_normal_force_at_top(double m, double v, double R, double g) {
    return m * (v * v / R - g);  /* can be negative = fall off */
}
double loop_the_loop_normal_force_at_bottom(double m, double v, double R, double g) {
    return m * (v * v / R + g);
}

/* ================================================================
 * L6: Incline-pulley combination
 *
 * Knowledge point: Combined constraints require solving coupled
 * equations. m1 on incline connected via pulley to hanging m2.
 *
 * Equation: m2*g - T = m2*a  (vertical)
 *           T - m1*g*sin(theta) - mu*m1*g*cos(theta) = m1*a  (incline)
 * -> a = (m2 - m1*(sin(theta) + mu*cos(theta))) * g / (m1 + m2)
 * ================================================================ */

void incline_pulley_system(double m1_on_incline, double m2_hanging,
                           double theta, double mu, double g,
                           double *acceleration, double *tension) {
    double M = m1_on_incline + m2_hanging;
    double f_incline = m1_on_incline * (sin(theta) + mu * cos(theta));
    double a = (m2_hanging - f_incline) * g / M;
    *acceleration = a;
    *tension = m2_hanging * (g - a);
}
int incline_pulley_direction(double m1, double m2, double theta,
                             double mu_s, double g) {
    double force_m2 = m2 * g;
    double force_m1_down = m1 * g * sin(theta);
    double friction_limit = mu_s * m1 * g * cos(theta);
    if (fabs(force_m2 - force_m1_down) <= friction_limit) return 0;  /* static */
    return (force_m2 > force_m1_down) ? 1 : -1;  /* 1=m1 up, -1=m1 down */
}

/* ================================================================
 * L6: Banked curves
 *
 * Knowledge point: On a frictionless banked curve, the normal
 * force provides the centripetal component.
 * tan(theta) = v^2/(gR) -> theta_ideal = arctan(v^2/(gR))
 *
 * With friction (mu):
 * v_max = sqrt(gR * (tan(theta) + mu) / (1 - mu*tan(theta)))
 * v_min = sqrt(gR * (tan(theta) - mu) / (1 + mu*tan(theta)))
 * ================================================================ */

double banked_curve_ideal_angle(double v, double R, double g) {
    double tan_theta = v * v / (g * R);
    return atan(tan_theta);
}
double max_safe_speed_banked(double R, double theta, double mu, double g) {
    double tan_t = tan(theta);
    double denom = 1.0 - mu * tan_t;
    if (denom <= 0.0) return 1e300;  /* no slip possible */
    return sqrt(g * R * (tan_t + mu) / denom);
}
double min_safe_speed_banked(double R, double theta, double mu, double g) {
    double tan_t = tan(theta);
    double denom = 1.0 + mu * tan_t;
    double num = tan_t - mu;
    if (num < 0.0) return 0.0;  /* friction holds even at rest */
    return sqrt(g * R * num / denom);
}

/* ================================================================
 * L4: Effective gravity in accelerating frames
 * g_eff = g_true - a_frame
 * In an elevator: apparent weight = m*(g + a)  (a>0 upward)
 * ================================================================ */

Vec3 effective_gravity_vec(Vec3 a_frame, Vec3 g_true) {
    return vec3_sub(g_true, a_frame);
}
double apparent_weight_in_elevator(double m, double a_elevator, double g) {
    return m * (g + a_elevator);
}

/* ================================================================
 * L6: Simple pendulum
 *
 * Knowledge point:
 * Small-angle: T = 2*pi*sqrt(L/g), omega = sqrt(g/L), theta(t) = theta0*cos(omega*t)
 * Large-angle: T = 2*pi*sqrt(L/g) * (1 + theta0^2/16 + 11*theta0^4/3072 + ...)
 * Full equation: d^2(theta)/dt^2 = -(g/L)*sin(theta)
 *
 * Course: MIT 8.012 Ch.10, Goldstein Ch.6
 * ================================================================ */

double pendulum_period_small_angle(double L, double g) {
    return 2.0 * M_PI * sqrt(L / g);
}
double pendulum_frequency(double L, double g) {
    return sqrt(g / L);
}
double pendulum_period_large_angle(double L, double theta0, double g) {
    /* Series expansion to 4th order:
     * T = T0 * (1 + theta0^2/16 + 11*theta0^4/3072 + 173*theta0^6/737280)
     */
    double T0 = pendulum_period_small_angle(L, g);
    double t2 = theta0 * theta0;
    double t4 = t2 * t2;
    double t6 = t4 * t2;
    double factor = 1.0 + t2/16.0 + 11.0*t4/3072.0 + 173.0*t6/737280.0;
    return T0 * factor;
}
double pendulum_angular_acceleration(double theta, double L, double g) {
    return -(g / L) * sin(theta);
}

/* ================================================================
 * L4: D'Alembert's principle
 *
 * Knowledge point: The virtual work of constraint forces is zero.
 * For a system in static/dynamic equilibrium with holonomic constraints:
 * sum_i (F_applied_i + F_constraint_i) . delta_r_i = 0
 * where delta_r_i are virtual displacements compatible with constraints.
 *
 * This is the foundation for Lagrangian mechanics.
 * ================================================================ */

double dalembert_virtual_work(int n_particles, Vec3 applied_forces[],
                              Vec3 virtual_displacements[]) {
    double delta_W = 0.0;
    for (int i = 0; i < n_particles; i++) {
        delta_W += vec3_dot(applied_forces[i], virtual_displacements[i]);
    }
    return delta_W;
}
